File: solutions/346.moving-average-from-data-stream/moving-average-from-data-stream.js
/**
 * Initialize your data structure here.
 * @param {number} size
 */
var MovingAverage = function(size) {
  this.size = size
  this.arr = []
}

/** 
* @param {number} val
* @return {number}
*/
MovingAverage.prototype.next = function(val) {
  this.arr.push(val)
  if (this.arr.length > this.size) {
    this.arr.shift()
    return this.arr.reduce((acc, e) => acc + e, 0) / this.size
  } else {
    return this.arr.reduce((acc, e) => acc + e, 0) / this.arr.length
  }
}

/** 
* Your MovingAverage object will be instantiated and called as such:
* var obj = Object.create(MovingAverage).createNew(size)
* var param_1 = obj.next(val)
*/

File: solutions/346.moving-average-from-data-stream/moving-average-from-data-stream.py
from collections import deque


class MovingAverage:

    def __init__(self, size):
        """
        Initialize your data structure here.
        :type size: int
        """
        self.size = size
        self.q = deque()

    def next(self, val):
        """
        :type val: int
        :rtype: float
        """
        self.q.append(val)

        if len(self.q) == self.size + 1:
            self.q.popleft()

        return sum(self.q) / len(self.q)


# Your MovingAverage object will be instantiated and called as such:
# obj = MovingAverage(size)
# jkj jj param_1 = obj.next(val)
